### DISCUSSION TITLE
How to download/install a bottle for the old version supporting Mojave?

### DISCUSSION THREAD

**Discussion by volo-droid** [Homebrew/discussions#2799] [Everyday usage] [answered]
Is there a way to download or install one of the previous versions of `scrcpy` formula bottles that supports Mojave? 
On Mojave the latest version of `scrcpy` formula is built from the sources, and installing a `rust` dependency takes ages to complete. GitHub packages contain previous version bottles that have binaries for Mojave. For example:
https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fscrcpy/9705995?tag=1.19

How could I install or extract those binaries?

**Comment by SMillerDev** [comment]
Since Mojave is no longer supported by Apple we dropped support for it too, this is why it's building from source. The best solution would be to update your OS so you can use bottles again.

If you want the latest version of scrcpy I'm afraid you'll always have to build from source on Mojave since the latest dependencies are no longer available for Mojave either. If you're fine installing any version you could check when the latest mojave bottle was available and download that version from the GitHub package registry. How to do that I'm not sure, [it's some magic with docker formats](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)

**Reply by volo-droid** [reply]
@SMillerDev thanks for quick the answer! Is there any reason why `brew` doesn't have a feature to install an old version from the GitHub registry?  (i.e. nobody requested it, or it goes against the philosophy of brew etc). I mean, the binaries are there so I'd assume it's not a big difference to fetch a specific version instead of the latest one.

**Reply by carlocab** [reply]
It's mostly against our philosophy -- we're a rolling release package manager. Trying to do what you're doing is a bit fragile and easily causes breakage for other formulae you have installed, which is another reason why we don't give users ways to do this easily. Note that Homebrew is built on Git, though, so knowing Git should be all you need to do what you want to do. (I'll refrain from spelling out the steps further for reasons already discussed.)

**Reply by fmillion-mnsu** [reply]
Out of curiosity, has there ever been any thought given to "locking" the last versions of packages currently supported before deprecating an older OS? For example, "if you want prebuilt bottles, this is _the_ version you  get from now on...if you want to go newer, you're on your own and you'll be building from source". I'd imagine something like Git branching could do this - admittedly I haven't looked into how the repos are set up yet, but was just a passing thought.

There's lots of reasons people might be stuck on an older OS - unsupported machine, some other app that isn't and won't be updated, a major feature the user depends on was changed in a breaking way in the new OS, etc. Apple of course deprecates their support pretty quickly, which is why I was thinking of "ok, we were supported up until today, so unless you add a flag like `--newest` or something, you get _these specific bottles_...and that's it, we won't be building/hosting any new ones." A lot of times these older machines and OSes are being kept around for purpose-specific needs and "the latest version" isn't always a necessity.

I can understand the potential cost issue of hosting binary blobs for older OS'es, but that might be where some kind of community support might need to step in.

**Reply by SMillerDev** [reply]
It's hard maintaining just the OS versions Apple supports with the small team Homebrew has, adding more OS versions that are not supported by Apple is not something that would be worth the effort in my book.

**Reply by volo-droid** [reply]
> I can understand the potential cost issue of hosting binary blobs for older OS'es

They already host those binaries. Try to run the following:
```
brew install oras
oras pull ghcr.io/homebrew/core/hello:2.10
```
and you'll end up with all the binaries from Monterey down to Yosemite. But to properly support the dependency mgmt. on older OS versions Homebrew would need to perform regular testing on those platforms and I think no one really wants to go down that rabbit hole.

**Comment by volo-droid** [comment, marked as answer]
In the end I was able to do it with `oras`. 
The following comments from one of `brew` maintainers provide valuable info on how bottles are stored and can be accessed: https://github.com/oras-project/oras/issues/237#issuecomment-815250008

**Reply by fuji246** [reply]
Could you share the command? thanks!

**Reply by volo-droid** [reply]
@fuji246 have a look at the comment here: https://github.com/oras-project/oras/issues/237#issuecomment-816964703
```
oras pull ghcr.io/homebrew/core/hello:2.10
```

**Reply by mahrud** [reply]
Do you know what is the syntax for versioned formulae like `gcc@12`? I'm specifically looking to see if there are any linux `tbb@2020` bottles are still accessible on github, but the `@` causes issues:
```
$ oras repo tags ghcr.io/homebrew/core/tbb@2020
Error: "ghcr.io/homebrew/core/tbb@2020": invalid reference: invalid digest "2020": invalid checksum digest format
Please make sure the provided reference is in the form of <registry>/<repo>[:tag|@digest]
```

edit: AH I think I found the answer:
```
  def self.image_formula_name(formula_name)
    # Invalid docker name characters:
    # - `/` makes sense because we already use it to separate repository/formula.
    # - `x` makes sense because we already use it in `Formulary`.
    formula_name.tr("@", "/")
                .tr("+", "x")
  end
```
